### ISSUE TITLE
BufferedStream.readline strips delimiter - makes it impossible to discover EOF

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2170]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2170|
|Type|defect|
|Created|2006-10-11 20:35:55Z|

This can lead to infinite loops and stuff.  In particular, wsgi apps seems to get stuck (ie, trac on pyramid).

<details><summary>Searchable metadata</summary>

```
trac-id__2170 2170
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1160598955000000 1160598955000000
changetime__1160882895000000 1160882895000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|foom set status to `closed`|
|-|-|

(In [18460]) Fix wsgi.input's readline, readlines, and __iter__ methods. In detail:

 * Remove BufferedStream.readline maxLength argument.
 * Add size argument to wsgi.input's readline(). This is not part of the WSGI spec, but code
   such as python stdlib cgi.py require it anyways. ([#11006](https://github.com/twisted/twisted/issues/11006))
 * Add BufferedStream.readline size argument, to enable above.

 * Fix wsgi.readlines() and wsgi.readline() to not add extraneous delimiters. ([[#2170](https://github.com/twisted/twisted/issues/2170)](https://github.com/twisted/twisted/issues/2170))
 * Make BufferedStream.readline return delimiter as part of the result, to enable above.

 * Add an __iter__ method to wsgi.input, which is required by the spec. ([[#2166](https://github.com/twisted/twisted/issues/2166)](https://github.com/twisted/twisted/issues/2166))

 * Change fileupload's use of BufferedStream.readline to conform with above changes.

Also, add some docstrings.

Author: mkerrin, jknight
Reviewer: exarkun
Merges branch: readline-1451
Fixes [#11006](https://github.com/twisted/twisted/issues/11006)
Fixes #2166
Fixes #2170

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
